dali/operators/math/expressions/broadcasting.cc

```diff
@@ -274,33 +274,97 @@ void ExpandToNDims(TensorShape<> &sh, int ndim) {
   sh = shape_cat(TensorShape<>(std::vector<int64_t>(ndim - sh.sample_dim(), 1)), sh);
 }
 
-void SimplifyShapesForBroadcasting(TensorShape<>& lhs, TensorShape<> &rhs) {
-  // First, if needed expand dimensions
-  int full_ndim = std::max(lhs.sample_dim(), rhs.sample_dim());
-  if (lhs.sample_dim() != rhs.sample_dim()) {
-    ExpandToNDims(lhs, full_ndim);
-    ExpandToNDims(rhs, full_ndim);
-  }
+void SimplifyShapesForBroadcasting(span<TensorShape<> *> shapes) {
+  int n = shapes.size();
+  SmallVector<TensorShape<>, 3> outs;
+  outs.resize(n);
+
+  int ndim = shapes[0]->size();
+  for (int i = 1; i < n; i++)
+    if (static_cast<int>(shapes[i]->size()) > ndim)
+      ndim = shapes[i]->size();
+
+  auto get = [&](int shape, int dim) -> int64_t {
+    auto &s = *shapes[shape];
+    dim -= ndim - s.size();  // add leading unit dims
+    if (dim < 0)
+      return 1;  // implicit unit dim
+    return s[dim];
+  };
+
+  auto should_skip = [&](int d) {
+    for (int i = 0; i < n; i++)
+      if (get(i, d) != 1)
+        return false;
+    return true;
+  };
+
+  SmallVector<int64_t, 3> volumes;
+  volumes.resize(n, 1);
+
+  int group_start = 0;
+
+  auto can_collapse = [&](int d) {
```

Comment:
Where is any kind of equality check on the matching dimensions? Or do we drop those, and assume we already checked it?

Reply:
I think we assume just that.

Reply:
We try to simplify shapes only after we've checked that they are compatible.